Lib/test/_test_multiprocessing.py

```diff
@@ -4472,17 +4473,34 @@ def check_semaphore_tracker_death(self, signum, should_die):
         # bpo-31310: if the semaphore tracker process has died, it should
         # be restarted implicitly.
         from multiprocessing.semaphore_tracker import _semaphore_tracker
-        _semaphore_tracker.ensure_running()
         pid = _semaphore_tracker._pid
+        if pid:
+            os.kill(pid, signal.SIGKILL)
+            time.sleep(0.5)  # give it time to die
+        old_stderr = sys.stderr
+        r, w = os.pipe()
+        try:
+            sys.stderr = open(w, "bw")
+            with warnings.catch_warnings(record=True) as all_warn:
+                _semaphore_tracker.ensure_running()
+            pid = _semaphore_tracker._pid
+            # Wait until we receive the PONG from the child, indicating that
+            # the signal handlers have been registered. See bpo-33613 for more
+            # information.
+            _semaphore_tracker._send("PING", "")
+            with open(r, "rb") as pipe:
+                data = pipe.readline()
```

Comment:
Are we okay with this potentially hanging indefinitely if something goes wrong?

Reply:
Commit d075473 implements a way to fail if reading takes too long.